### ISSUE TITLE
broken manifest download from Artifactory Container Registry API

### ISSUE THREAD

**Issue by suvl** [argoproj-labs/argocd-image-updater#306]
We are using the Artifactory Container Registry as our docker image CRI and since the last update, I believe the one that introduced OCI support, it broke the manifest download from a list of image tags.

These are the logs for version `0.4.2`:

```text
time="2021-11-29T12:19:08Z" level=debug msg="Considering this image for update" alias=boimage application=some-project/backoffice image_name=some-project/backoffice image_tag=20211129T1123-dev registry=docker.my-company.com
time="2021-11-29T12:19:08Z" level=debug msg="Using no version constraint when looking for a new tag" alias=boimage application=some-project/backoffice image_name=some-project/backoffice image_tag=20211129T1123-dev registry=docker.my-company.com
time="2021-11-29T12:19:09Z" level=debug msg="Cache hit for some-project/backoffice:20211112T1117-dev"
time="2021-11-29T12:19:09Z" level=debug msg="Cache hit for some-project/backoffice:20211112T1120-dev"
time="2021-11-29T12:19:09Z" level=debug msg="Cache hit for some-project/backoffice:20211112T1147-dev"
time="2021-11-29T12:19:09Z" level=debug msg="Cache hit for some-project/backoffice:20211112T1155-dev"
time="2021-11-29T12:19:09Z" level=debug msg="Cache hit for some-project/backoffice:20211112T1158-dev"
time="2021-11-29T12:19:09Z" level=debug msg="Cache hit for some-project/backoffice:20211112T1224-dev"
time="2021-11-29T12:19:09Z" level=debug msg="Cache hit for some-project/backoffice:20211112T1408-dev"
time="2021-11-29T12:19:09Z" level=debug msg="Cache hit for some-project/backoffice:20211112T1415-dev"
time="2021-11-29T12:19:09Z" level=debug msg="Cache hit for some-project/backoffice:20211119T1659-dev"
time="2021-11-29T12:19:09Z" level=debug msg="Cache hit for some-project/backoffice:20211123T1449-dev"
time="2021-11-29T12:19:09Z" level=debug msg="Cache hit for some-project/backoffice:20211124T1144-dev"
time="2021-11-29T12:19:09Z" level=debug msg="Cache hit for some-project/backoffice:20211125T1704-dev"
time="2021-11-29T12:19:09Z" level=debug msg="Cache hit for some-project/backoffice:20211125T1723-dev"
time="2021-11-29T12:19:09Z" level=debug msg="Cache hit for some-project/backoffice:20211125T1740-dev"
time="2021-11-29T12:19:09Z" level=debug msg="Cache hit for some-project/backoffice:20211125T1749-dev"
time="2021-11-29T12:19:09Z" level=debug msg="Cache hit for some-project/backoffice:20211125T1809-dev"
time="2021-11-29T12:19:09Z" level=debug msg="Cache hit for some-project/backoffice:20211126T1037-dev"
time="2021-11-29T12:19:09Z" level=debug msg="Cache hit for some-project/backoffice:20211126T1106-dev"
time="2021-11-29T12:19:09Z" level=debug msg="Cache hit for some-project/backoffice:20211126T1108-dev"
time="2021-11-29T12:19:09Z" level=debug msg="Cache hit for some-project/backoffice:20211126T1123-dev"
time="2021-11-29T12:19:09Z" level=debug msg="Cache hit for some-project/backoffice:20211126T1140-dev"
time="2021-11-29T12:19:09Z" level=debug msg="Cache hit for some-project/backoffice:20211126T1146-dev"
time="2021-11-29T12:19:09Z" level=debug msg="Cache hit for some-project/backoffice:20211126T1157-dev"
time="2021-11-29T12:19:09Z" level=debug msg="Cache hit for some-project/backoffice:20211126T1427-dev"
time="2021-11-29T12:19:09Z" level=debug msg="Cache hit for some-project/backoffice:20211129T1010-dev"
time="2021-11-29T12:19:09Z" level=debug msg="Cache hit for some-project/backoffice:20211129T1123-dev"
time="2021-11-29T12:19:09Z" level=debug msg="found 26 from 26 tags eligible for consideration" image="docker.my-company.com/some-project/backoffice:20211129T1123-dev"
time="2021-11-29T12:19:09Z" level=debug msg="Image 'docker.my-company.com/some-project/backoffice:20211129T1123-dev' already on latest allowed version" alias=boimage application=some-project/backoffice image_name=some-project/backoffice image_tag=20211129T1123-dev registry=docker.my-company.com
time="2021-11-29T12:19:09Z" level=info msg="Processing results: applications=1 images_considered=1 images_skipped=0 images_updated=0 errors=0"
```

Now for the new `0.5.0` version, same configurations:

```text
time="2021-11-29T12:13:27Z" level=debug msg="Considering this image for update" alias=boimage application=some-project/backoffice image_name=some-project/backoffice image_tag=20211129T1123-dev registry=docker.my-company.com
time="2021-11-29T12:13:27Z" level=debug msg="Using no version constraint when looking for a new tag" alias=boimage application=some-project-backoffice image_name=some-project/backoffice image_tag=20211129T1123-dev registry=docker.my-company.com
time="2021-11-29T12:13:28Z" level=error msg="Error fetching metadata for some-project/backoffice:20211112T1120-dev - neither V1 or V2 or OCI manifest returned by registry: unknown: Bad Request"
time="2021-11-29T12:13:29Z" level=error msg="Error fetching metadata for some-project/backoffice:20211125T1749-dev - neither V1 or V2 or OCI manifest returned by registry: unknown: Bad Request"
time="2021-11-29T12:13:29Z" level=error msg="Error fetching metadata for some-project/backoffice:20211123T1449-dev - neither V1 or V2 or OCI manifest returned by registry: unknown: Bad Request"
time="2021-11-29T12:13:29Z" level=error msg="Error fetching metadata for some-project/backoffice:20211112T1147-dev - neither V1 or V2 or OCI manifest returned by registry: unknown: Bad Request"
time="2021-11-29T12:13:29Z" level=error msg="Error fetching metadata for some-project/backoffice:20211112T1155-dev - neither V1 or V2 or OCI manifest returned by registry: unknown: Bad Request"
time="2021-11-29T12:13:29Z" level=error msg="Error fetching metadata for some-project/backoffice:20211125T1723-dev - neither V1 or V2 or OCI manifest returned by registry: unknown: Bad Request"
time="2021-11-29T12:13:29Z" level=error msg="Error fetching metadata for some-project/backoffice:20211112T1158-dev - neither V1 or V2 or OCI manifest returned by registry: unknown: Bad Request"
time="2021-11-29T12:13:29Z" level=error msg="Error fetching metadata for some-project/backoffice:20211125T1740-dev - neither V1 or V2 or OCI manifest returned by registry: unknown: Bad Request"
time="2021-11-29T12:13:29Z" level=error msg="Error fetching metadata for some-project/backoffice:20211126T1123-dev - neither V1 or V2 or OCI manifest returned by registry: unknown: Bad Request"
time="2021-11-29T12:13:29Z" level=error msg="Error fetching metadata for some-project/backoffice:20211124T1144-dev - neither V1 or V2 or OCI manifest returned by registry: unknown: Bad Request"
time="2021-11-29T12:13:29Z" level=error msg="Error fetching metadata for some-project/backoffice:20211126T1037-dev - neither V1 or V2 or OCI manifest returned by registry: unknown: Bad Request"
time="2021-11-29T12:13:29Z" level=error msg="Error fetching metadata for some-project/backoffice:20211112T1117-dev - neither V1 or V2 or OCI manifest returned by registry: unknown: Bad Request"
time="2021-11-29T12:13:29Z" level=error msg="Error fetching metadata for some-project/backoffice:20211112T1415-dev - neither V1 or V2 or OCI manifest returned by registry: unknown: Bad Request"
time="2021-11-29T12:13:29Z" level=error msg="Error fetching metadata for some-project/backoffice:20211125T1704-dev - neither V1 or V2 or OCI manifest returned by registry: unknown: Bad Request"
time="2021-11-29T12:13:29Z" level=error msg="Error fetching metadata for some-project/backoffice:20211126T1427-dev - neither V1 or V2 or OCI manifest returned by registry: unknown: Bad Request"
time="2021-11-29T12:13:29Z" level=error msg="Error fetching metadata for some-project/backoffice:20211126T1108-dev - neither V1 or V2 or OCI manifest returned by registry: unknown: Bad Request"
time="2021-11-29T12:13:29Z" level=error msg="Error fetching metadata for some-project/backoffice:20211126T1146-dev - neither V1 or V2 or OCI manifest returned by registry: unknown: Bad Request"
time="2021-11-29T12:13:29Z" level=error msg="Error fetching metadata for some-project/backoffice:20211125T1809-dev - neither V1 or V2 or OCI manifest returned by registry: unknown: Bad Request"
time="2021-11-29T12:13:29Z" level=error msg="Error fetching metadata for some-project/backoffice:20211119T1659-dev - neither V1 or V2 or OCI manifest returned by registry: unknown: Bad Request"
time="2021-11-29T12:13:29Z" level=error msg="Error fetching metadata for some-project/backoffice:20211129T1010-dev - neither V1 or V2 or OCI manifest returned by registry: unknown: Bad Request"
time="2021-11-29T12:13:29Z" level=error msg="Error fetching metadata for some-project/backoffice:20211126T1157-dev - neither V1 or V2 or OCI manifest returned by registry: unknown: Bad Request"
time="2021-11-29T12:13:29Z" level=error msg="Error fetching metadata for some-project/backoffice:20211126T1140-dev - neither V1 or V2 or OCI manifest returned by registry: unknown: Bad Request"
time="2021-11-29T12:13:29Z" level=error msg="Error fetching metadata for some-project/backoffice:20211126T1106-dev - neither V1 or V2 or OCI manifest returned by registry: unknown: Bad Request"
time="2021-11-29T12:13:29Z" level=error msg="Error fetching metadata for some-project/backoffice:20211112T1224-dev - neither V1 or V2 or OCI manifest returned by registry: unknown: Bad Request"
time="2021-11-29T12:13:29Z" level=error msg="Error fetching metadata for some-project/backoffice:20211112T1408-dev - neither V1 or V2 or OCI manifest returned by registry: unknown: Bad Request"
time="2021-11-29T12:13:29Z" level=error msg="Error fetching metadata for some-project/backoffice:20211129T1123-dev - neither V1 or V2 or OCI manifest returned by registry: unknown: Bad Request"
time="2021-11-29T12:13:29Z" level=debug msg="Image 'docker.my-company.com/some-project/backoffice:20211129T1123-dev' already on latest allowed version" alias=boimage application=some-project-backoffice image_name=some-project/backoffice image_tag=20211129T1123-dev registry=docker.my-company.com
time="2021-11-29T12:13:29Z" level=info msg="Processing results: applications=1 images_considered=1 images_skipped=0 images_updated=0 errors=0"
```

Install is being done by helm, we've simply updated the chart and started noticing the problem. I'm afraid that steps to reproduce involve setting up an artifactory with the proper reverse proxy configurations and then configuring argo-cd image updater to use it as docker container registry. We cannot confirm if this happens on other container registries as we use solely Artifactory.

Any steps you might require to debug this, just let me know. We're available to help.

**Comment by jannfis** [comment]
Hi, sorry for the late reply. This sounds bad. Could you eventually increase the log level to `trace` to gain some more information?

I unfortunately don't have access to an Artifactory registry to test.

**Comment by mariobris** [comment]
Hi,
we are running into the same issue with Artifactory after upgrade argocd image updater.
```
time="2021-12-21T10:31:26Z" level=trace msg="released semaphore and terminated waitgroup"
time="2021-12-21T10:31:26Z" level=trace msg="acquired metadata semaphore"
time="2021-12-21T10:31:26Z" level=trace msg="Getting manifest for image someimage:211213.3 (operation 41/42)"
time="2021-12-21T10:31:26Z" level=trace msg="https://artifactory.url.com/v2/someimage/manifests/211213.1"
time="2021-12-21T10:31:26Z" level=error msg="Error fetching metadata for someimage:211005.4 - neither V1 or V2 or OCI manifest returned by registry: unknown: Bad Request"
time="2021-12-21T10:31:26Z" level=trace msg="released semaphore and terminated waitgroup"
time="2021-12-21T10:31:26Z" level=trace msg="acquired metadata semaphore"
time="2021-12-21T10:31:26Z" level=trace msg="Getting manifest for image someimage:211216.8 (operation 42/42)"
time="2021-12-21T10:31:26Z" level=trace msg="https://artifactory.url.com/v2/someimage/manifests/211213.3"
time="2021-12-21T10:31:26Z" level=error msg="Error fetching metadata for someimage:211007.1 - neither V1 or V2 or OCI manifest returned by registry: unknown: Bad Request"
time="2021-12-21T10:31:26Z" level=trace msg="released semaphore and terminated waitgroup"
time="2021-12-21T10:31:26Z" level=trace msg="acquired metadata semaphore"
time="2021-12-21T10:31:26Z" level=trace msg="https://artifactory.url.com/v2/someimage/manifests/211216.8"
time="2021-12-21T10:31:26Z" level=error msg="Error fetching metadata for someimage:210830.3 - neither V1 or V2 or OCI manifest returned by registry: unknown: Bad Request"
time="2021-12-21T10:31:26Z" level=trace msg="released semaphore and terminated waitgroup"
time="2021-12-21T10:31:26Z" level=error msg="Error fetching metadata for someimage:211012.1 - neither V1 or V2 or OCI manifest returned by registry: unknown: Bad Request"
time="2021-12-21T10:31:26Z" level=trace msg="released semaphore and terminated waitgroup"
time="2021-12-21T10:31:26Z" level=error msg="Error fetching metadata for someimage:210818.4 - neither V1 or V2 or OCI manifest returned by registry: unknown: Bad Request"
time="2021-12-21T10:31:26Z" level=error msg="Error fetching metadata for someimage:210818.5 - neither V1 or V2 or OCI manifest returned by registry: unknown: Bad Request"
time="2021-12-21T10:31:26Z" level=trace msg="released semaphore and terminated waitgroup"
time="2021-12-21T10:31:26Z" level=trace msg="released semaphore and terminated waitgroup"
```
Here are trace logs (made up AF url and image name)

**Comment by mariobris** [comment]
I'll try to summarize what I found so far in case someone will know how to check or fix it.
I did pull manifest.json of docker image from Artifactory to get some info..
```
{
   "schemaVersion": 2,
   "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
   "config": {
      "mediaType": "application/vnd.docker.container.image.v1+json",
      "size": 7441,
      "digest": "sha256:1f1b173cd86e332aed4bcbccec43a7bfd8a647a7a5a71787c1edccfb37649f87"
   },
   "layers": [
      {
         "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
         "size": 2711827,
         "digest": "sha256:552d1f2373af9bfe12033568ebbfb0ccbb0de11279f9a415a29207e264d7f4d9"
```

Some investigation of the output
- go package for schema2 is [github.com/distribution/distribution/v3/manifest/schema2](https://github.com/argoproj-labs/argocd-image-updater/blob/release-0.11/pkg/registry/client.go#L17)
- `distribution/manifest/schema2` seems to support it [mediaTypes](https://github.com/distribution/distribution/blob/6affafd1f030087d88f88841bf66a8abe2bf4d24/manifest/schema2/manifest.go#L15-L18)
- documentation for [manifest-v2.md](https://github.com/distribution/distribution/blob/1cdeff259b9d1793d034f725062cea559bf0a69f/docs/spec/manifest-v2-2.md) looks ok as well.

Argocd image updater error section 
- failing on [neither V1 or V1 or OCI manifest](https://github.com/argoproj-labs/argocd-image-updater/blob/master/pkg/registry/registry.go#L140)
- we take value for above error from [client.go](https://github.com/argoproj-labs/argocd-image-updater/blob/master/pkg/registry/client.go#L154)

Do not know what value it gets or expect to be yet. 

The change happend due to [OCI image compatibility #236](https://github.com/argoproj-labs/argocd-image-updater/issues/236) and we use different packages for [schemas](https://github.com/argoproj-labs/argocd-image-updater/pull/249/files#diff-acbc1e889c3e91003a88f18e21c3073b64937801b20f663fcf2601540b9e3cdaL17-R17) which might be the cause of failing on Artifactory registry with argocd image version 0.11.0.

I've tried to upload OCI image as well however that one is failing as well.
```
{
    "schemaVersion": 2,
    "config": {
        "mediaType": "application/vnd.oci.image.config.v1+json",
        "digest": "sha256:5bf0541b0568065c1b051b66f8d850aaf06cee995047a20924c0e183e33bf0f8",
        "size": 8169
    },
    "layers": [
        {
            "mediaType": "application/vnd.oci.image.layer.v1.tar+gzip",
            "digest": "sha256:44071ebb44952f778ec6e5efd5a3f1ed444e5ef3e6fa651fc4d6de966d92579c",
            "size": 2802353
```

I can get both images via `docker pull`, `podman pull` or `curl`. Anybody have an idea how to check whether the issue is on Artifactory side or argocd image updater ?

**Comment by jannfis** [comment]
So I finally came around to dig into this issue and created myself a trial version on JFrog's SaaS offering. I was able to reproduce the bug, and fortunately also to fix it.

The root cause is that we also request to request Docker's legacy V1 schema [here](https://github.com/argoproj-labs/argocd-image-updater/blob/8664d1445f0c879d7dbe90e83461e7e0d09171cf/pkg/registry/client.go#L159), but `schema1.SchemaVersion.MediaType` actually expands to an empty string and Artifactory doesn't like that.

I will submit a PR and release a v0.11.1 with this fix included soon.
